i3lock.c

```diff
@@ -1132,6 +1151,131 @@ static const struct raw_pixel_format raw_fmt_bgr = {3, 2, 1, 0};
 static const struct raw_pixel_format raw_fmt_bgrx = {4, 2, 1, 0};
 static const struct raw_pixel_format raw_fmt_xbgr = {4, 3, 2, 1};
 
+static cairo_surface_t *read_gif_image(const char *image_path) {
+    int err;
+    int width, stride, height;
+    int bg_idx;
+    uint32_t bg_color;
+    ColorMapObject *cmap;
+    GraphicsControlBlock gc;
+
+    /* Open and load a GIF file */
+    GifFileType *gif = DGifOpenFileName(image_path, &err);
+    if (!gif) {
+        fprintf(stderr, "Could not open GIF file, err=%d\n", err);
+        return NULL;
+    }
+    if (DGifSlurp(gif) != GIF_OK) {
+        printf("Could not read the GIF image\n");
+        goto read_gif_image_clean;
+    }
+
+    /* Load canvas properties */
+    width = gif->SWidth;
+    height = gif->SHeight;
+    cmap = gif->SColorMap;
+    bg_idx = gif->SBackGroundColor;
+    GifColorType *cmap_bg_rgb = cmap->Colors + bg_idx;
+    bg_color = 0 | cmap_bg_rgb->Blue | cmap_bg_rgb->Green<<8 | cmap_bg_rgb->Red<<16;
+    stride = cairo_format_stride_for_width(CAIRO_FORMAT_RGB24, width)/sizeof(uint32_t);
+    if (stride < 0) {
+        fprintf(stderr, "Invalid stride\n");
+        goto read_gif_image_clean;
+    }
+
+    /* Read the images into the RGB buffers */
+    // TODO: check C char specification, whether to just use int
+    gif_img = malloc(gif->ImageCount * sizeof(struct gif));
+    gif_img_count = gif->ImageCount;
+    if (!gif_img) {
+        fprintf(stderr, "Can't allocate memory for GIF image buffers\n");
+        goto read_gif_image_clean;
+    }
+    uint32_t *data_prev = NULL;
+    for (SavedImage *pimg = gif->SavedImages; pimg < gif->SavedImages + gif->ImageCount; ++pimg) {
+        int idx = pimg - gif->SavedImages;
+        /* Create image surface */
+        gif_img[idx].img = cairo_image_surface_create(CAIRO_FORMAT_RGB24, width, height);
+        if (cairo_surface_status(gif_img[idx].img) != CAIRO_STATUS_SUCCESS) {
+            fprintf(stderr, "Could not create surface: %s\n",
+                    cairo_status_to_string(cairo_surface_status(gif_img[idx].img)));
+            free(gif_img);
+            gif_img = NULL;
+            goto read_gif_image_clean;
+        }
+        cairo_surface_flush(gif_img[idx].img);
+        uint32_t *data = (uint32_t *)cairo_image_surface_get_data(gif_img[idx].img);
+
+        // TODO: Chek L,R,W,H && color map?
+        /* Find Graphics Control extension and delay */
+        for (int iext = 0; iext < pimg->ExtensionBlockCount; ++iext) {
+            ExtensionBlock *pext = pimg->ExtensionBlocks + iext;
+            switch (pext->Function) {
+                case GRAPHICS_EXT_FUNC_CODE:
+                    DGifExtensionToGCB(pext->ByteCount, pext->Bytes, &gc);
+                    gif_img[idx].delay_sec = gc.DelayTime*0.01;
```

Comment:
Spaces around multiplication operator please. Also, why are we multiplying by 0.01?

Reply:
`DelayTime` is in [ms], so we need to convert it to [s]. Added a comment about it.